tools/functional-tester/etcd-runner/main.go

```diff
@@ -212,6 +214,97 @@ func runRacer(eps []string, round int) {
 	doRounds(rcs, round)
 }
 
+func runWatcher(eps []string) {
+	type watcher struct {
+		prefix string
+		c      *clientv3.Client
+		wc     clientv3.WatchChan
+	}
+	prefixes := 36
+	watchPerKey := 10
+	ctx := context.Background()
+	prefixKeysMap := make(map[string][]string)
+	ws := make([]watcher, prefixes*watchPerKey)
+
+	var (
+		wg  sync.WaitGroup
+		pr  *clientv3.PutResponse
+		err error
+	)
+
+	const (
+		chars = "abcdefghijklmnopqrstuvwxyz0123456789"
+	)
+
+	wg.Add(len(ws))
+	for i := 0; i < len(ws); {
+		for _, prefix := range chars {
+			go func(w *watcher, i int, prefix rune) {
+				defer wg.Done()
+				w.c = randClient(eps)
+				w.prefix = string(prefix)
+				w.wc = w.c.Watch(ctx, w.prefix, clientv3.WithPrefix())
+			}(&ws[i], i, prefix)
+			i++
+		}
+	}
+	wg.Wait()
+
+	done := make(chan bool)
+	go func(strlen int, runningTime float64, reqRate int, done chan bool) {
+		client := randClient(eps)
+		defer client.Close()
+		rand.Seed(time.Now().UTC().UnixNano())
+		result := make([]byte, strlen)
+		start := time.Now()
+		reqStart := time.Now()
+		reqCount := 0
+		for time.Since(start).Seconds() < runningTime {
+			for i := 0; i < strlen; i++ {
+				result[i] = chars[rand.Intn(len(chars))]
+			}
+			key := string(result)
+			pr, err = client.Put(ctx, key, key)
+			if err != nil {
+				log.Fatalf("Error occured while trying to key: %v, value : %v to kv store.", key, key)
+			} else {
+				prefix := key[0:1]
+				vals := prefixKeysMap[prefix]
+				prefixKeysMap[prefix] = append(vals, key)
+				reqCount++
+			}
+			if reqCount >= reqRate*5 {
+				sleepTime := 5 - time.Since(reqStart).Seconds()
+				if sleepTime > 0 {
+					time.Sleep(time.Duration(sleepTime))
+				}
+				reqCount = 0
+				reqStart = time.Now()
+			}
+		}
+		done <- true
+	}(10, 30, 20, done)
+	<-done
+
+	for _, w := range ws {
```

Comment:
@xiang90 Before raising this PR i had tried making it concurrent, but it was causing values mismatches, hence i had removed it. May be will try it out using locks.